Adithya impl avg months pledged frontend

adithya-6101 opened the pull request:
# Description
Implements Job Posting Page Analytics component featuring a horizontal bar chart displaying "Average Number of Months Pledged by Role." The component provides interactive filtering by dates and roles, displays data labels on each bar, includes hover tooltips showing exact average months pledged, and maintains responsive design across different screen sizes. The chart sorts roles from highest to lowest average pledged months, addressing the analytics dashboard requirements for job posting insights.

## Related PRS (if any):
This frontend PR is related to the https://github.com/OneCommunityGlobal/HGNRest/pull/1654  - Backend PR.
To test this frontend PR you need to checkout the https://github.com/OneCommunityGlobal/HGNRest/pull/1654 backend PR.

## Main changes explained:
- Created MonthsPledgedChart.jsx for horizontal bar chart component with highest-to-lowest sorting and filtering logic  
- Created MonthsPledgedChart.css for responsive bar chart styling
- Implemented dynamic fetching of data from backend `/analytics/months-pledged` API  
- Implemented "Average Number of Months Pledged by Role" title, data labels on bars, and hover tooltips showing exact averages  
- Added Date filter (All/Custom Range) and Role filter (All + individual roles) with default "All" selections  

## How to test:
1. Check out the current branch  
2. Run `npm install` and `npm start` to run the app locally  
3. Clear site data/cache  
4. Log in as admin user  
5. Go to http://localhost:5173/analytics/months-pledged
6. Verify the horizontal bar chart displays average months per role  
7. Test filters: change date range or select specific roles → chart updates correctly

## Screenshots or videos of changes:

https://github.com/user-attachments/assets/e72790c4-5cf4-4df3-ac53-0a6a54036b7b

adithya-6101 commented:
Hi @PipKcK, thanks for checking this. I think the issue might be that npm install wasn’t run after switching to the branch, so the required data didn’t load correctly on your frontend.

I’ve attached a screenshot showing the results for the same date filters on my setup, everything appears correctly on my end.

<img width="1890" height="681" alt="Screenshot 2025-08-14 212020" src="https://github.com/user-attachments/assets/48644976-444b-457f-82de-fa402aec110b" />

akshith312 reviewed (approved):
Getting the desired output.
Had to modify the .env file and update the REACT_APP_APIENDPOINT to http://localhost:4500/api

<img width="719" height="365" alt="image" src="https://github.com/user-attachments/assets/6c5e8a6a-fdf0-4f88-afab-3ac2f30f59bd" />
<img width="720" height="362" alt="image" src="https://github.com/user-attachments/assets/e7273db7-e318-45d7-91c7-1169fe93e37a" />

sourabhbagde reviewed (dismissed):
Checked out the branch and npm, logged in as admin. Set frontend .env to
REACT_APP_APIENDPOINT="http://localhost:4500/api"
(as per note about staging).

Navigated to http://localhost:5173/analytics/months-pledged, the horizontal bar chart renders and shows average months per role.

**Outcomes:**
- Role filter → working correctly
- Date range filter → updates chart as expected
- Reset button → restores defaults as expected

Overall, feature works as described with the local API endpoint (attached screenshots below)

Note: Is this chart also styled in dark mode? Because it was not reflecting in dark mode.

<img width="1440" height="900" alt="PR 3886+1654 date" src="https://github.com/user-attachments/assets/507e72fe-1929-4f62-9f82-ecdcacfdd5c5" />
<img width="1440" height="900" alt="PR 3886+1654 filter " src="https://github.com/user-attachments/assets/686e14d8-22ce-4fb8-a133-80e85be38de5" />
<img width="1440" height="900" alt="PR 3886+1654 reset " src="https://github.com/user-attachments/assets/fddf077b-c734-4cd7-a183-cb05b48f8b12" />
<img width="1440" height="900" alt="PR 3886+1654 dark" src="https://github.com/user-attachments/assets/6b53d536-cb71-428a-98ea-452cff0ff3d3" />

harshavarma29 reviewed (changes requested):
I have tested the PR locally using Admin/Test and Volunteer accounts. It is working as expected for all the scenarios tested.

However, I noticed that the UI is not changing in dark mode.

<img width="1886" height="842" alt="dark_mode_issue" src="https://github.com/user-attachments/assets/bb75430a-2a79-43bc-b3fc-b5f7015f09fa" />

khan-zoha reviewed (approved):
I reviewed PR 3886. I followed the instructions and ensured your frontend .env file is configured correctly. The horizontal bar chart displays average months per role and filter works as expected. Great work!

<img width="1344" height="695" alt="Screenshot 2025-09-13 at 5 33 51 PM" src="https://github.com/user-attachments/assets/b101a3bb-1275-46c1-a2a9-37902d204eb9" />

Aditya-gam reviewed (changes requested):
- Replaced the `REACT_APP_APIENDPOINT` with "http://localhost:4500/api" in the .env
- Navigated to http://localhost:5173/analytics/months-pledged and logged in with an admin account.
- Tested the functionalities of the page.
	- https://github.com/user-attachments/assets/66fcbd5f-3ae1-45ec-9e91-076d36e4ee54

- One issue I saw is that if I try to enter the start date > end date, I cannot do that if I start by entering the start date first.
	- https://github.com/user-attachments/assets/b04a1bd9-76f3-4a70-a39e-ce8f1c3ba92a

- But if I start with entering the end date first, i can enter start date > end date.
	-  https://github.com/user-attachments/assets/2f1196ac-459e-4d7d-9250-bb105204048a

	- It doesn't return any data.

laynet reviewed (changes requested):
I had some difficulty initially while trying to test this, not sure if that was my fault or not (i thought I was doing it correctly, and It worked when I restarted to double check) but i'll still report what happened in case this wasn't an error on my end but please disregard if it's not relevant.
I ran backend on correct branch, then followed the front end testing instructions. I went to the http://localhost:5173/analytics/months-pledged  page and I got a pop-up error that i wasn’t authorized to access the resource but the page was still displaying the ui and filters
![PR 3886 months pledged pop up](https://github.com/user-attachments/assets/2822577f-1271-42f7-8770-1c5ac70cc3cf)
I used the dates from the example video but no roles were available. I tried to type in a role and that didn’t work. I reset filters and tried random dates as well but there were never an roles available. I also tried setting the end date first and then the start date. (I'm assuming because of the pop-up error)
![PR 3886 no roles](https://github.com/user-attachments/assets/f51b2f1d-d9a4-469e-902f-f32f9c0517ad)
I went back to the dashboard and got an error, and also when I tried to see my profile. I was trying to verify that I was still logged in as an admin and got an infinity skeleton load on my profile page.
![PR 3886 profile error](https://github.com/user-attachments/assets/0062479d-b35d-45a3-87db-83df8caeb80f)
![PR 3886 profile page infinity load](https://github.com/user-attachments/assets/9d05a70d-8444-42ef-be3e-1fbe58769103)
I cleared my site data, logged out and re-logged back in as an admin, then when I went back to the months-pledged route I was able to see the chart. I verified the ui is working as expected with the dates from the example video.
![PR 3886 ui verified](https://github.com/user-attachments/assets/481a50ab-3560-4934-ace1-6b4e6ba2e161)
When I tried other dates there were no roles available, when I deleted my start date but kept my end date the chart was still displaying data. This doesn't seem to be the intended behavior, which is why I marked this for request changes (assuming that the trouble seeing this page as an admin was my error)
![PR 3886 no start date](https://github.com/user-attachments/assets/e0b2d7e7-a08e-40cd-85b1-79003cfdc726)
When I reset the filter and tried again I was able to see data displaying as expected.

Prem203 commented:
Reviewed the PR and the graph is working as expected in both light and dark modes. The date filter works and start date cannot be after end date is also handled. 

The filter by role dropdown and reset filters work as expected.

<img width="1893" height="951" alt="image" src="https://github.com/user-attachments/assets/d30329d6-9bd7-4240-a76b-1a96c3f35911" />

<img width="1867" height="725" alt="image" src="https://github.com/user-attachments/assets/b1429955-a8d3-4489-98da-8ec6fd642c81" />

<img width="1894" height="957" alt="image" src="https://github.com/user-attachments/assets/f2427b99-cec9-466c-a8a4-fe0f224458de" />

harshavarma29 reviewed (changes requested):
Hi Adithya,

Thank you for addressing the comments and implementing dark mode. I noticed a minor issue, the graph boundaries are still displayed in light mode.

<img width="1900" height="835" alt="dark_mode_not_working" src="https://github.com/user-attachments/assets/37128730-3942-428f-b309-43363790c165" />

adithya-6101 commented:
Hi Aryan,

Thanks a lot for reviewing! I’ve addressed your suggestions:

Dark mode styling – Updated the chart and surrounding components to reflect dark mode consistently.

Date validation – Start date cannot be after the end date, regardless of input order.

Single date selection – Clarified behavior so only one date selection shows a valid range; invalid selections are restricted.

Unauthorized access handling – Added clearer error messages and redirect behavior for unauthorized access.

Could you please review the changes and approve the PR if everything looks good?

Thanks!

adithya-6101 commented:
> Hi Adithya,
> 
> Thank you for addressing the comments and implementing dark mode. I noticed a minor issue, the graph boundaries are still displayed in light mode.
> 
> <img alt="dark_mode_not_working" width="1923" height="840" src="https://private-user-images.githubusercontent.com/46283874/491999841-d1651113-02eb-4676-a755-97096dc6032e.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTkzNTQ1MzksIm5iZiI6MTc1OTM1NDIzOSwicGF0aCI6Ii80NjI4Mzg3NC80OTE5OTk4NDEtZDE2NTExMTMtMDJlYi00Njc2LWE3NTUtOTcwOTZkYzYwMzJlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEwMDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMDAxVDIxMzAzOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJlZWZkZTEyZmY2NThkNDVhZTM0ZWY2MzFmMDhhN2NhNDNmN2IzZDJiNDdjY2E2YjU0MTg1MTYwMTMwYTY5N2MmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.2UsbAz_ipmK-OofyTg7AEUSmcHEh4PsZO9lLG7YL_2E">

Hi Harsha,

Thanks for the feedback! I’ve fixed the graph boundaries for dark mode. Could you please review and approve if everything looks good?

Thanks!

one-community commented:
Thank you all, merging!
